src/consensus/block_producer.rs

```diff
@@ -289,47 +201,130 @@ where
+        let (block_hash, num_txs) = self.produce_slices(slot, parent, None).await?;
+        Ok(((slot, block_hash), num_txs))
+    }
 
-                (payload, left)
+    async fn produce_slices(
+        &self,
+        slot: Slot,
+        parent: BlockId,
+        mut parent_ready_receiver: Option<&mut oneshot::Receiver<BlockId>>,
+    ) -> Result<(Hash, usize)> {
+        // only start the DELTA_BLOCK timer when ParentReady event is seen
+        let mut duration_left = match parent_ready_receiver {
+            None => self.delta_block,
+            Some(_) => Duration::MAX,
+        };
+        let mut txs_in_block = 0;
+
+        for slice_index in SliceIndex::all() {
+            let slice_parent = slice_index.is_first().then_some(parent);
+            // HACK: use of `MAX_SLICES_PER_BLOCK / 2`, maybe introduce `MAX_SLICES_FIRST_SLOT`
+            // HACK: use of `SliceIndex::inner()`, maybe introduce `is_last_before_parent_ready()`
+            let min_time_for_slice = if parent_ready_receiver.is_some()
+                && slice_index.inner() == MAX_SLICES_PER_BLOCK / 2
+            {
+                Duration::MAX
             } else {
-                produce_slice_payload(&self.txs_receiver, None, duration_left).await
+                Duration::ZERO
             };
+            let (preempt_sender, preempt_receiver) = oneshot::channel();
+            let produce_slice_future = produce_slice_payload(
+                &self.txs_receiver,
+                slice_parent,
+                min_time_for_slice,
+                preempt_receiver,
+            );
+            pin!(produce_slice_future);
+
+            let (payload, new_duration_left, txs_in_slice) =
+                if let Some(ref mut parent_rcv) = parent_ready_receiver {
+                    // if we have not yet received the ParentReady event,
+                    // wait for it concurrently while producing the next slice
+                    tokio::select! {
+                        res = &mut produce_slice_future => {
+                            let (payload, _time_elapsed, txs_in_slice) = res;
+                            // ParentReady event still not seen, do not start DELTA_BLOCK timer yet
+                            (payload, Duration::MAX, txs_in_slice)
+                        }
+                        res = parent_rcv => {
+                            parent_ready_receiver = None;
+                            // Got ParentReady event while producing slice.
+                            // It's a NOP if we have been using the same parent as before.
+                            let start = Instant::now();
+                            let (new_slot, new_hash) = res.expect("sender dropped");
+                            let (payload, _time_elapsed, txs_in_slice) = tokio::select! {
+                                () = tokio::time::sleep(self.delta_first_slice) => {
+                                    preempt_sender.send(()).unwrap();
```

Comment:
Hmm... I think this `unwrap` is not safe.

`produce_slice_payload` might return without receiving a value on the channel i.e. dropping the channel.  And if this `send` were to happen after the drop, it will panic.